mesonbuild/dependencies/blas_lapack.py

```diff
@@ -0,0 +1,136 @@
+# Copyright 2013-2020 The Meson development team
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from pathlib import Path
+import functools
+import os
+import re
+import typing as T
+
+from .. import mlog
+from .. import mesonlib
+
+from .base import DependencyMethods, SystemDependency
+from .base import DependencyException
+from .cmake import CMakeDependency
+from .factory import factory_methods
+
+if T.TYPE_CHECKING:
+    from ..environment import Environment, MachineChoice
+    from .factory import DependencyGenerator
+
+
+# TODO: how to select BLAS interface layer (LP64, ILP64)?
+#       LP64 is the 32-bit interface (confusingly)
+#       ILP64 is the 64-bit interface
+#
+#       OpenBLAS library names possible:
+#           - libopenblas
+#           - libopenblas64_  # are these suffixes NumPy-specific or not? If so, don't support it?
+#           - libopenblas_ilp64
+#
+# MKL docs: https://www.intel.com/content/www/us/en/develop/documentation/onemkl-windows-developer-guide/top/linking-your-application-with-onemkl/linking-in-detail/linking-with-interface-libraries/using-the-ilp64-interface-vs-lp64-interface.html
+#
+# TODO: do we care here about whether OpenBLAS is compiled with pthreads or OpenMP?
+#       MKL has separate .pc files for this (_seq vs. _iomp), OpenBLAS does not
+#
+# TODO: what about blas/lapack vs. cblas/clapack?
+#       See http://nicolas.limare.net/pro/notes/2014/10/31_cblas_clapack_lapacke/ for details
+#
+# NOTE: we ignore static libraries for now
+
+
+@factory_methods({DependencyMethods.PKGCONFIG, DependencyMethods.CMAKE, DependencyMethods.SYSTEM})
+def openblas_factory(env: 'Environment', for_machine: 'MachineChoice',
+                  kwargs: T.Dict[str, T.Any],
+                  methods: T.List[DependencyMethods]) -> T.List['DependencyGenerator']:
+    candidates: T.List['DependencyGenerator'] = []
+
+    if DependencyMethods.CMAKE in methods:
```

Comment:
You probably want to yield the pkg-config one first, right?

Instead of defining your own `func foo_factory` you can assign `foo_factory = DependencyFactory(name, methods_list, ...)`, which is a simplified API that just creates a factory returning one of each method in the listed order, using the default or overridden name parameters. Example from misc.py:

```python
cups_factory = DependencyFactory(
    'cups',
    [DependencyMethods.PKGCONFIG, DependencyMethods.CONFIG_TOOL, DependencyMethods.EXTRAFRAMEWORK, DependencyMethods.CMAKE],
    configtool_class=CupsDependencyConfigTool,
    cmake_name='Cups',
)
```